File: fixtures/scale/dart/large/lib/ciphers/runningkey/runningkey_cipher.dart
import '../../core/cipher.dart';
import '../../util/alphabet.dart';
import 'runningkey_key.dart';

/// Vigenere with a long passage as the key stream instead of a short word.
class RunningKeyCipher implements Cipher {
  final RunningKeyKey key;

  RunningKeyCipher(this.key);

  @override
  String name() => 'runningkey';

  @override
  String encrypt(String plaintext) => _transform(cleanAlphabet(plaintext), true);

  @override
  String decrypt(String ciphertext) => _transform(cleanAlphabet(ciphertext), false);

  String _transform(String text, bool forward) {
    final sb = StringBuffer();
    for (var i = 0; i < text.length; i++) {
      final x = indexOf(text[i]);
      final k = indexOf(key.keyCharAt(i));
      sb.write(charAt(forward ? x + k : x - k));
    }
    return sb.toString();
  }
}

File: fixtures/scale/scala/large/src/cryptobench/ciphers/runningkey/RunningKeyCipher.scala
package cryptobench.ciphers.runningkey

import cryptobench.core.Cipher
import cryptobench.util.Alphabet

/** Vigenere with a long passage as the key stream instead of a short word. */
final class RunningKeyCipher(key: RunningKeyKey) extends Cipher {

  override def name: String = "runningkey"

  override def encrypt(plaintext: String): String = transform(Alphabet.clean(plaintext), forward = true)

  override def decrypt(ciphertext: String): String = transform(Alphabet.clean(ciphertext), forward = false)

  private def transform(text: String, forward: Boolean): String = {
    val sb = new StringBuilder
    for (i <- text.indices) {
      val x = Alphabet.indexOf(text.charAt(i))
      val k = Alphabet.indexOf(key.keyCharAt(i))
      sb.append(Alphabet.charAt(if (forward) x + k else x - k))
    }
    sb.toString
  }
}
